File: app/src/main/java/nu/huw/clarity/model/Perspective.java
package nu.huw.clarity.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.IdRes;
import android.support.annotation.StyleRes;

/**
 * Holds data about a perspective which can be used for organisation
 */
public class Perspective extends Base {

  public static final Parcelable.Creator<Perspective> CREATOR =
      new Parcelable.Creator<Perspective>() {
        public Perspective createFromParcel(Parcel in) {

          return new Perspective(in);
        }

        public Perspective[] newArray(int size) {

          return new Perspective[size];
        }
      };
  @ColorRes
  public int colorID;
  @ColorRes
  public int colorStateListID;
  @DrawableRes
  public int cursorDrawable;
  @DrawableRes
  public int icon;
  @IdRes
  public int menuID;
  @StyleRes
  public int themeID;
  public String filterDuration;
  public String filterFlagged;
  public String filterStatus;
  public String collation;
  public String name;
  public String sort;
  public String value;
  public String viewMode;

  public Perspective() {
  }

  private Perspective(Parcel in) {

    super(in);

    // res
    colorID = in.readInt();
    colorStateListID = in.readInt();
    icon = in.readInt();
    menuID = in.readInt();
    themeID = in.readInt();

    // properties
    filterDuration = in.readString();
    filterFlagged = in.readString();
    filterStatus = in.readString();
    collation = in.readString();
    name = in.readString();
    sort = in.readString();
    value = in.readString();
    viewMode = in.readString();
  }

  @Override
  public void writeToParcel(Parcel out, int flags) {

    super.writeToParcel(out, flags);

    // res
    out.writeInt(colorID);
    out.writeInt(colorStateListID);
    out.writeInt(icon);
    out.writeInt(menuID);
    out.writeInt(themeID);

    // properties
    out.writeString(filterDuration);
    out.writeString(filterFlagged);
    out.writeString(filterStatus);
    out.writeString(collation);
    out.writeString(name);
    out.writeString(sort);
    out.writeString(value);
    out.writeString(viewMode);
  }

  /**
   * Compare perspectives based on their IDs rather than object references
   */
  @Override
  public boolean equals(Object obj) {
    return obj instanceof Perspective && id.equals(((Perspective) obj).id);
  }
}

File: app/src/main/java/nu/huw/clarity/model/Perspective.kt
package nu.huw.clarity.model

import android.arch.persistence.room.Entity
import android.arch.persistence.room.Ignore
import android.arch.persistence.room.TypeConverters
import nu.huw.clarity.R
import nu.huw.clarity.db.Converters
import org.threeten.bp.LocalDateTime
import java.util.*

@Entity
@TypeConverters(Converters::class)
class Perspective(id: ID = ID(), dateAdded: LocalDateTime = LocalDateTime.now(), dateModified: LocalDateTime = LocalDateTime.now(), var filterDuration: DurationFilterState = DurationFilterState.NONE, var filterFlagged: FlaggedFilterState = FlaggedFilterState.NONE, var filterStatus: StatusFilterState = StatusFilterState.NONE, var collation: CollationState = CollationState.NONE, var name: String = "", var sort: SortState = SortState.NONE, var data: String? = null, var viewMode: ViewModeState = ViewModeState.CONTEXT, var iconState: PerspectiveIconState = PerspectiveIconState.INBOX, var colorState: PerspectiveColorState = PerspectiveColorState.BLUE_GREY) : Base(id, dateAdded, dateModified) {
    @Ignore constructor() : this(ID())

    @Ignore val random: Random = Random()

    @Ignore
    var icon = {
        when (iconState) {
            PerspectiveIconState.FLAGGED -> R.drawable.ic_flag_orange
            PerspectiveIconState.CONTEXTS -> R.drawable.ic_contexts_purple
            PerspectiveIconState.PROJECTS -> R.drawable.ic_projects_blue
            PerspectiveIconState.INBOX -> R.drawable.ic_inbox_bluegrey
            PerspectiveIconState.FORECAST -> R.drawable.ic_forecast_red
            PerspectiveIconState.NEARBY -> R.drawable.ic_nearby_green
        }
    }

    @Ignore
    var colorID = {
        when (colorState) {
            PerspectiveColorState.ORANGE -> R.color.primary_orange
            PerspectiveColorState.PURPLE -> R.color.primary
            PerspectiveColorState.BLUE -> R.color.primary_blue
            PerspectiveColorState.BLUE_GREY -> R.color.primary_blue_grey
            PerspectiveColorState.RED -> R.color.primary_red
            PerspectiveColorState.GREEN -> R.color.primary_green
        }
    }

    @Ignore
    var colorStateListID = {
        when (colorState) {
            PerspectiveColorState.ORANGE -> R.color.state_list_orange
            PerspectiveColorState.PURPLE -> R.color.state_list_purple
            PerspectiveColorState.BLUE -> R.color.state_list_blue
            PerspectiveColorState.BLUE_GREY -> R.color.state_list_blue_grey
            PerspectiveColorState.RED -> R.color.state_list_red
            PerspectiveColorState.GREEN -> R.color.state_list_green
        }
    }

    @Ignore
    var themeID = {
        when (colorState) {
            PerspectiveColorState.ORANGE -> R.style.AppTheme_Orange
            PerspectiveColorState.PURPLE -> R.style.AppTheme
            PerspectiveColorState.BLUE -> R.style.AppTheme_Blue
            PerspectiveColorState.BLUE_GREY -> R.style.AppTheme_BlueGrey
            PerspectiveColorState.RED -> R.style.AppTheme_Red
            PerspectiveColorState.GREEN -> R.style.AppTheme_Green
        }
    }

    @Ignore
    var menuID = {
        when (id.value) {
            "ProcessFlagged" -> R.id.menuitem_main_flagged
            "ProcessContexts" -> R.id.menuitem_main_context
            "ProcessProjects" -> R.id.menuitem_main_projects
            "ProcessInbox" -> R.id.menuitem_main_inbox
            "ProcessForecast" -> R.id.menuitem_main_forecast
            else -> random.nextInt()
        }
    }
}